### PR TITLE
Command execution tests

### PR DESCRIPTION
**Pull request by oxarbitrage**
I had been exploring the possibility of command line tests for https://github.com/ZcashFoundation/zebra/issues/677

This is work in progress and it is open to get feedback of the overall idea and for help/opinions in some specific problems.

The problem that worries me most is that when command like `start` or `seed` are tested i cant kill the spawned `zebrad` process even if the timeout option is working for exiting the tests, a zombie process remains. This is happening for example in `seed_no_args` and `start_no_args`. Any opinion on that will be great.

### CODE REVIEW

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: please call `wait` to check the exit status:
https://docs.rs/abscissa_core/0.5.2/abscissa_core/testing/process/struct.Process.html#method.wait

We need to call `wait` on every process we launch.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: On Unix-based operating systems (Linux, macOS), zombie processes stick around until the parent process calls `wait`:
https://docs.rs/abscissa_core/0.5.2/abscissa_core/testing/process/struct.Process.html#method.wait

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Question: What does the usage documentation for `zebrad start` say about its arguments?

Does it say anything about extra arguments?
Do we need to reject extra arguments for all our commands?

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Please call wait here

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Please call wait here

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Please call wait here

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Please call wait here

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Please call wait here

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks like a good start!

I think most of your issues can be fixed by calling `wait` on every child process.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: The `wait()` haves an error as we made the process time out. I added `cmd.wait().unwrap_err();` at https://github.com/ZcashFoundation/zebra/pull/690/commits/1d05af32b82f6ee49ffa50ef827953ba7b99364f however the process is still around. Maybe i am not calling it correctly ?

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs]: This can be too much maintenance if abscissa changes the strings too often. We can maybe just keep the `assert_failure()` without comparing the string in this cases(invalid argument, invalid flag, etc).

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: Its important to not drop this `_guard`, doing so will delete the temp dir that this command is going to run inside of.

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs:11]: I think the only way to effectively deduplicate these would be to use a macro

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: probably related to the guard that's being dropped

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]:
```suggestion
    let mut child = child.unwrap();
```

TIOLI: It's probably better to just shadow the variable here since we're never going to reuse the original binding to `child`

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: Whats wrong with this test case rn?

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: I think it would be okay to add an `info` event that announces that we're starting up for `start` and `seed`

cc @hdevalence

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Any extra argument gets rejected automatically except if there is an actual command with that name. In the case of `start` we know it it can work combined with `seed`. All others free arguments or flags should be rejected. I will research further and see if i can make this happen.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I think it should be an error however zebra is starting.

Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: More at https://github.com/ZcashFoundation/zebra/pull/690#discussion_r458405609

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I think that will be good.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Thanks, fixed this at https://github.com/ZcashFoundation/zebra/pull/690/commits/5f038d540267a53f6ae4e80756e5af9a2818f5f6

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: It was, thanks!

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: Why'd you switch to waiting on the inner `Child` here? This will not add context to the error report if something goes wrong, such as what the original command was.

Edit: OOH its because you didn't destructure the tuple, one sec, let me make a suggestion

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]:
```suggestion
    let (child, _guard) = get_child_single_arg("generate")?;
```

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Thanks.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Applied everywhere at https://github.com/ZcashFoundation/zebra/pull/690/commits/75bbaacfcb26c10edff2c7a4f44004b228b9b85a

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Resolved in previous commit.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I added these at https://github.com/ZcashFoundation/zebra/pull/690/commits/9d97cfb07052a44e0351904e7db87b9a0a6cef27

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I got confused as the issue says in one of the items: `Seed (might be combined with Start)`

This is not actually true, the `start` command accept any free additional argument however doing `zebrad start seed` will do nothing:

```
$ ./zebrad start seed
Jul 22 14:58:47.430  INFO zebrad::components::tracing: Initializing tracing endpoint
Jul 22 14:58:47.430  INFO zebrad::components::metrics: Initializing metrics endpoint
^C
$ 
```
This extra arguments are for overwriting configuration https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/commands/start.rs#L92 but it haves nothing to do with running 2 commands in one.

I fixed the `start_args` test case accordingly in https://github.com/ZcashFoundation/zebra/pull/690/commits/9d97cfb07052a44e0351904e7db87b9a0a6cef27

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/690#discussion_r458981570

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Removed all these at https://github.com/ZcashFoundation/zebra/pull/690/commits/38207cbe023d972fc9bc83a7deffad7126dadc2c

**Comment by yaahc** [comment]
Last I checked this is blocked on the port issue, lmk if there's anything I can do to help.

**Comment by oxarbitrage** [comment]
Yes, i dont have a clear idea on how we can go over this. Any help is welcome.

I know `$ cargo test -- --test-threads=1` is possible to run one test after the other. It is maybe possible to do this setup in the CI for this particular tests. However by what i saw in discord from discussions yesterday this will not fix the issue.

**Comment by yaahc** [comment]
@oxarbitrage I think the first step should be to test out moving the tests that use the port from their current file to separate files and test to see if rustc runs each test file one at a time or if it runs them in parallel. If it runs them in parallel check back in and I'll help you figure out what to try next.

**Comment by oxarbitrage** [comment]
Tried at https://github.com/ZcashFoundation/zebra/pull/690/commits/ee74ad877f47639705809a266a2d665990cb5d28 but same problem.

File: zebrad/tests/start_no_args.rs
Comment (yaahc) [inline zebrad/tests/start_no_args.rs]: TIOLI: At this point I don't think there's much value to having this broken out into its own function, I'd inline this code back into the `start_no_args` fn personally.

Comment (oxarbitrage) [reply zebrad/tests/start_no_args.rs]: Yea, this was actually just for testing the CI, i had the plan of making this functions callable from here(or just inline it) however as separating by now didn't make any difference this split can be reverted.

Comment (yaahc) [reply zebrad/tests/start_no_args.rs]: aah yea, makes sense. I'm gonna pull this down locally and play around with it a bit.

Comment (yaahc) [reply zebrad/tests/start_no_args.rs]: Okay so I got an okay initial solution I think. At least good enough to unblock this. I combined those three tests into a single test like so:

```rust
#[test]
fn serialized_tests() -> Result<()> {
    start_no_args()?;
    start_args()?;
    seed_no_args()?;

    Ok(())
}
```

This has the disadvantage that we will only see one item in the test output from `cargo` but I think that's okay for now.

Comment (yaahc) [reply zebrad/tests/start_no_args.rs]: gonna push a commit that adds this

**Review by yaahc — DISMISSED** [review]
okay, this looks good to me. Let's hope the tests pass! I've also gone ahead and re-requested a review from @teor2345 so we can get final approval and merge done on this if you're so inclined @oxarbitrage

File: zebrad/tests/start_no_args.rs
Comment (oxarbitrage) [reply zebrad/tests/start_no_args.rs]: You are the best!

**Comment by oxarbitrage** [comment]
Was about to push pretty much the same changes. Thanks! Fingers crossed now.

**Comment by yaahc** [comment]
@oxarbitrage it looks like this one's good to go, would you mind fixing the merge conflict and re-pushing?

**Review by yaahc — DISMISSED** [review]

**Comment by oxarbitrage** [comment]
Apologies @yaahc but i had to submit a new commit https://github.com/ZcashFoundation/zebra/pull/690/commits/62e4a817ff34694540499ab1ec429b5177b015e9. In most recent version it seems the behaviour of `start` changed a bit so now a second argument to the subcommand will not produce any output but the program is started. The behaviour from before(just ignoring the extra argument) and the one we have now(silently start the program) are both wrong imho but probably will be better to research and fix this in a new PR if you agree.

**Comment by dconnolly** [comment]
The gcloud build failure:

image

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: I don't think this test is worth it as is, I'd prefer to just remove this entirely rather than only testing that we can spawn and send a kill signal to a command without an error.

**Review by yaahc — CHANGES_REQUESTED** [review]

**Review by yaahc — DISMISSED** [review]

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Yea, agree. Thanks for making it.

File: zebra-test/src/command.rs
Comment (teor2345) [inline zebra-test/src/command.rs]:
```suggestion
    /// wrapper for `spawn` fn on `Command` that constructs informative error
```

File: zebra-test/src/command.rs
Comment (teor2345) [inline zebra-test/src/command.rs]:
```suggestion
    /// wrapper for `spawn` fn on `Command` that constructs informative error
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:11]: We could move this code into a utility function:

            .stdout(Stdio::piped())
             .stderr(Stdio::piped())
             .spawn2()
             .unwrap(),

**Review by teor2345 — APPROVED** [review]

**Review by teor2345 — COMMENTED** [review]
Thanks for writing these command tests.

They will help us catch bugs that break the zebrad commands.

The seed and start runs just check for failure. But they don't check for the difference between a panic and a kill. I opened #804 to check for the kill signal.

**Pull request merged**
